sympy/matrices/matrices.py

```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
         LUdecompositionFF
         LUsolve
         """
-        combined, p = self.LUdecomposition_Simple(iszerofunc=_iszero)
-        L = self.zeros(self.rows)
-        U = self.zeros(self.rows)
-        for i in range(self.rows):
-            for j in range(self.rows):
-                if i > j:
-                    L[i, j] = combined[i, j]
-                else:
-                    if i == j:
-                        L[i, i] = 1
-                    U[i, j] = combined[i, j]
+
+        combined, p = self.LUdecomposition_Simple(iszerofunc=iszerofunc,
+                                                  rankcheck=rankcheck)
+
+        L = self.zeros(combined.rows)
+        U = self.zeros(combined.rows, combined.cols)
+
+        # L is lower triangular self.rows x self.rows
+        # U is upper triangular self.rows x self.cols
+        # L has unit diagonal. For each column in combined, the subcolumn
+        # below the diagonal of combined is shared by L.
+        # If L has more columns than combined, then the remaining subcolumns
+        # below the diagonal of L are zero.
+        # The upper triangular portion of L and combined are equal.
+
+        max_index = min(self.rows, self.cols)
+
+        # Set diagonal entry to 1, and subdiagonal entries of L to
+        # subdiagonal entries of combined
+        for col_i in range(max_index):
+            L[col_i, col_i] = 1
+            for row_i in range(col_i + 1, L.rows):
+                L[row_i, col_i] = combined[row_i, col_i]
+                row_i += 1
+
+        # Fill in the remaining diagonal entries of L
+        for col_i in range(max_index, L.cols):
+            L[col_i, col_i] = 1
+
+        # Set upper triangular portion of U to upper triangular portion of
+        # combined
+        for row_i in range(max_index):
+            for col_i in range(row_i, U.cols):
+                U[row_i, col_i] = combined[row_i, col_i]
+
         return L, U, p
 
-    def LUdecomposition_Simple(self, iszerofunc=_iszero):
-        """Returns A comprised of L, U (L's diag entries are 1) and
-        p which is the list of the row swaps (in order).
+    def LUdecomposition_Simple(self,
+                               iszerofunc=_iszero,
+                               rankcheck=False):
+        """Compute an lu decomposition of m x n matrix A, where P*A = L*U
+        L is m x m lower triangular with unit diagonal
+        U is m x n upper triangular
+        P is an m x m permutation matrix
```

Comment:
Leave blank lines for new paragraphs.  This also might benefit from being a list.

For example
```
Compute an lu decomposition of m x n matrix A, where P*A = L*U

    * L is m x m lower triangular with unit diagonal
    * U is m x n upper triangular
    * P is an m x m permutation matrix

Returns an m x n ...
```

Sphinx will create nice HTML if you format this way.